// eslint-plugin-postcss-modules/dist/settings.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const path_1 = tslib_1.__importDefault(require("path"));
const types_1 = require("./types");
const getSettings = ({ settings: { "postcss-modules": settings }, }) => settings || {};
class Settings {
    constructor(context) {
        this.contextDir = path_1.default.dirname(context.getFilename());
        const settings = getSettings(context);
        this.postcssConfigDir = settings.postcssConfigDir || this.contextDir;
        this.baseDir = settings.baseDir || "";
        this.camelCase =
            typeof settings.camelCase !== "undefined" ? settings.camelCase : false;
        this.defaultScope = settings.defaultScope || types_1.LocalScope;
        this.include = settings.include || "**/*.css";
        this.exclude = settings.exclude || "**/node_modules/**/*";
    }
    resolveFile(filename) {
        const baseDir = filename.startsWith(".") ? this.contextDir : this.baseDir;
        return path_1.default.resolve(baseDir, filename);
    }
}
exports.default = Settings;
