// eslint-plugin-postcss-modules/dist/rules/no-undef-class.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const common_1 = require("./common");
exports.default = common_1.createRule({
    description: "Ensures that any referenced class is exported by css files.",
    messages: {
        undefinedClassName: "{{ className }} does not exist in {{ filename }}",
    },
    create: (context) => {
        const cache = new common_1.Cache(context);
        return {
            Program: () => {
                common_1.Cache.clear();
            },
            ImportDeclaration: (node) => {
                const result = cache.processImportDeclaration(node);
                if (result === null) {
                    return;
                }
                const { filename, classes, explicitImports } = result;
                explicitImports.forEach((node) => {
                    const className = node.imported.name;
                    if (!classes.has(className)) {
                        context.report({
                            messageId: "undefinedClassName",
                            node,
                            data: {
                                filename,
                                className,
                            },
                        });
                    }
                });
            },
            MemberExpression: (node) => {
                const result = cache.processMemberExpression(node);
                if (result === null) {
                    return;
                }
                const { filename, className, classes } = result;
                if (!classes.has(className) && !className.startsWith("_")) {
                    context.report({
                        messageId: "undefinedClassName",
                        node,
                        data: {
                            filename,
                            className,
                        },
                    });
                }
            },
        };
    },
});
