// eslint-plugin-postcss-modules/dist/rules/common/create-rule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createRule = void 0;
exports.createRule = ({ description, messages, addlProps, create, }) => ({
    meta: {
        docs: {
            description,
            recommended: true,
        },
        messages,
        schema: [
            {
                type: "object",
                properties: {
                    baseDir: {
                        description: "Base directory for resolving 'absolute' imports",
                        type: "string",
                    },
                    camelCase: {
                        description: "How classes are exported. See the documentation for css-loader: https://github.com/webpack-contrib/css-loader/tree/v2.1.1#camelcase",
                        oneOf: [
                            { type: "boolean" },
                            {
                                enum: [
                                    "asIs",
                                    "camelCase",
                                    "camelCaseOnly",
                                    "dashes",
                                    "dashesOnly",
                                    "only",
                                ],
                            },
                        ],
                    },
                    defaultScope: {
                        description: "The default scope of classes that are not explicitly scoped: local, global, or pure... I'm not sure what pure is; it's not documented.",
                        enum: ["local", "global", "pure"],
                    },
                    include: {
                        description: "Anymatch describing what files to parse.",
                        type: "any",
                    },
                    exclude: {
                        description: "Anymatch describing what files to exclude from parsing.",
                        type: "any",
                    },
                    ...addlProps,
                },
                additionalProperties: false,
            },
        ],
    },
    create,
});
